test/apicoverage-webhook/webhook_handlers.go

```diff
@@ -0,0 +1,171 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"container/list"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"reflect"
+
+	"github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	"github.com/knative/test-infra/tools/webhook-apicoverage/coveragecalculator"
+	"github.com/knative/test-infra/tools/webhook-apicoverage/resourcetree"
+	"github.com/knative/test-infra/tools/webhook-apicoverage/view"
+	"go.uber.org/zap"
+	"k8s.io/api/admission/v1beta1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+)
+
+const (
+	// ResourceQueryParam query param name to provide the resource.
+	ResourceQueryParam = "resource"
+)
+
+var (
+	scheme = runtime.NewScheme()
+	codecs = serializer.NewCodecFactory(scheme)
+)
+
+// APICoverageRecorder type contains resource tree to record API coverage for resources.
+type APICoverageRecorder struct {
+	Decoder runtime.Decoder
+	Logger *zap.SugaredLogger
+	resourceForest resourcetree.ResourceForest
+}
+
+func (a *APICoverageRecorder) init() {
+	a.resourceForest = resourcetree.ResourceForest {
+		Version: "v1alpha1",
+		ConnectedNodes: make(map[string]*list.List),
+		TopLevelTrees: make(map[string]resourcetree.ResourceTree),
+	}
+	a.addResourceTree("Service", reflect.TypeOf(v1alpha1.Service{}))
+	a.addResourceTree("Configuration", reflect.TypeOf(v1alpha1.Configuration{}))
+	a.addResourceTree("Route", reflect.TypeOf(v1alpha1.Route{}))
+	a.addResourceTree("Revision", reflect.TypeOf(v1alpha1.Revision{}))
+}
+
+func (a *APICoverageRecorder) addResourceTree(resource string, t reflect.Type) {
+	tree := resourcetree.ResourceTree{
+		ResourceName: resource,
+		Forest: &a.resourceForest,
+	}
+	tree.BuildResourceTree(t)
+	a.resourceForest.TopLevelTrees[resource] = tree
+}
+
+func (a *APICoverageRecorder) recordCoverage(review *v1beta1.AdmissionReview) {
+	var (
+		v reflect.Value
+		resource string
+		err error
+	)
+
+	switch schemaType := review.Request.Kind.Kind; schemaType {
+	case "Service":
+		resource = "Service"
+		svc := v1alpha1.Service{}
+		err = json.Unmarshal(review.Request.Object.Raw, &svc)
+		// Even if err is encountered we would just reflect on an empty object.
+		v = reflect.ValueOf(svc)
+	case "Configuration":
+		resource = "Configuration"
+		conf := v1alpha1.Configuration{}
+		err = json.Unmarshal(review.Request.Object.Raw, &conf)
+		v = reflect.ValueOf(conf)
+	case "Route":
+		resource = "Configuration"
+		route := v1alpha1.Route{}
+		err = json.Unmarshal(review.Request.Object.Raw, &route)
+		v = reflect.ValueOf(route)
+	case "Revision":
+		resource = "Revision"
+		revision := v1alpha1.Revision{}
+		err = json.Unmarshal(review.Request.Object.Raw, &revision)
+		v = reflect.ValueOf(revision)
+	default:
+		return
+	}
```

Comment:
This should use the map passed in elsewhere in the webhook (that is it's express purpose).